web/packages/teleport/src/ThemeProvider.tsx

```diff
@@ -0,0 +1,125 @@
+/**
+ * Teleport
+ * Copyright (C) 2024 Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, useEffect, ReactNode } from 'react';
+import { StaticThemeProvider } from 'design/ThemeProvider';
+
+import { Theme as ThemePreference } from 'gen-proto-ts/teleport/userpreferences/v1/theme_pb';
+
+import { bblpTheme, lightTheme, darkTheme, Theme } from 'design/theme';
+
+import cfg from 'teleport/config';
+import { storageService, KeysEnum } from 'teleport/services/storageService';
+
+export const ThemeProvider = (props: { children?: ReactNode }) => {
+  const [themePreference, setThemePreference] = useState<ThemePreference>(
+    storageService.getThemePreference()
+  );
+
+  useEffect(() => {
+    storageService.subscribe(receiveMessage);
+
+    function receiveMessage(event) {
+      const { key, newValue } = event;
+
+      if (!newValue || key !== KeysEnum.USER_PREFERENCES) {
+        return;
+      }
+
+      const preferences = JSON.parse(newValue);
+      if (
+        preferences.theme !== ThemePreference.UNSPECIFIED &&
+        preferences.theme !== themePreference
+      ) {
+        setThemePreference(preferences.theme);
+      }
+    }
+
+    // Cleanup on unmount
+    return function unsubscribe() {
+      storageService.unsubscribe(receiveMessage);
+    };
+  }, [themePreference]);
+
+  const customThemes = {
+    bblp: bblpTheme,
+  };
+
+  // If no props.theme is defined, use the custom theme instead of the user preference theme.
+  let theme: Theme;
+  if (customThemes[cfg.customTheme]) {
+    theme = customThemes[cfg.customTheme];
+  } else {
+    theme = themePreferenceToTheme(themePreference);
+  }
+
+  return (
+    <StaticThemeProvider theme={theme}>{props.children}</StaticThemeProvider>
+  );
+};
+
+function themePreferenceToTheme(themePreference: ThemePreference): Theme {
+  if (themePreference === ThemePreference.UNSPECIFIED) {
+    return getPrefersDark() ? lightTheme : darkTheme;
+  }
+  return themePreference === ThemePreference.LIGHT ? lightTheme : darkTheme;
+}
+
+// because unspecified can exist but only used as a fallback and not an option,
+// we need to get the current/next themes with getPrefersDark in mind.
+// TODO (avatus) when we add user settings page, we can add a Theme.SYSTEM option
+// and remove the checks for unspecified
```

Comment:
Actually since this is an exported function, how about make this a proper JSDoc?